### ISSUE TITLE
20260924_Train, validate, then test three binary links_balanced accuracy, auc, sensitivity, specificity, accuracy

### ISSUE THREAD

**Issue by guzhiling** [USC-STAT517-2026-Fall/inclass-labs#4]
Use the familiar credit-card example to compare **logit, probit, and complementary log-log** regression. Every group fits all three models with the same predictors: student status, balance, and income. The event is **default = Yes**.

Our workflow is **60% training → 20% validation → 20% test**:

| Subset | What you do with it |
|---|---|
| Training | Estimate the coefficients of each model. |
| Validation | Choose a separate classification threshold for each link. |
| Test | Compare the three fixed classification rules on new observations. |

All three links must use the same split within your group. Groups use different fixed random seeds so we can see how results vary across splits. Keep your assigned seed and model formula throughout the lab.

## Group assignments

Group1: @Kent-Robinson, @swalser01, @blurissa

Group2: @BrooklynTyner, @cjdow, @jmames17

Group3: @Peyton-Math, @jcarp333, @zhuang-jy

Group4: @jess-l-b, @zoemelton, @maggieleporin

Group5: @nmayers517, @jth27, @joshyoung803

Group6: @Isabellhaines, @asearly60

Choose someone to run the code, someone to check the data flow, and someone to record the interpretation. The two-person group can combine roles. The six student groups are separate from the three data subsets.

## 1. Make the 60/20/20 split — 5 minutes

Use [Default.csv](data/Default.csv). In RStudio, work in `Code/class10/`, with the file inside its `data/` folder. If downloading the lab separately, create the same `data/Default.csv` arrangement. Copy the code blocks into an R script and run them in order. Only the `pROC` package is needed.

Change `group` to your group number. The supplied loop shuffles customers separately within the two outcome categories, keeping the default rate similar across subsets. For these data, rounding gives exactly **6,000 training, 2,000 validation, and 2,000 test customers**.

```r
# Install once if needed: install.packages("pROC")
library(pROC)

group <- 1L  # CHANGE to your group number: 1, 2, 3, 4, 5, or 6.
set.seed(2400 + group)

credit <- read.csv("data/Default.csv", stringsAsFactors = TRUE)
credit$default <- factor(credit$default, levels = c("No", "Yes"))
credit$student <- factor(credit$student, levels = c("No", "Yes"))

# Each customer belongs to exactly one subset.
split <- character(nrow(credit))
for (outcome in levels(credit$default)) {
  rows <- sample(which(credit$default == outcome))
  n <- length(rows)
  n.train <- round(0.60 * n)
  n.valid <- round(0.20 * n)
  split[rows] <- rep(c("Training", "Validation", "Test"),
                    times = c(n.train, n.valid, n - n.train - n.valid))
}

train <- credit[split == "Training", ]
valid <- credit[split == "Validation", ]
test <- credit[split == "Test", ]
c(Training = nrow(train), Validation = nrow(valid), Test = nrow(test))
```

**Discuss:** Why should each link use the same training customers? Which subset will choose the threshold, and which will assess its performance?

## 2. Fit the three links on training data — 5 minutes

Only `train` enters these fits. A binomial model with the factor levels `No`, `Yes` models the probability of **Yes**. Keep the same predictors under every link.

```r
logit.fit <- glm(default ~ student + balance + income,
                 data = train, family = binomial(link = "logit"))
probit.fit <- glm(default ~ student + balance + income,
                  data = train, family = binomial(link = "probit"))
cloglog.fit <- glm(default ~ student + balance + income,
                   data = train, family = binomial(link = "cloglog"))

summary(logit.fit)
summary(probit.fit)
summary(cloglog.fit)
```

**Check:** Identify where the training data and link function enter each call. Why would comparing the raw coefficient magnitudes across links be misleading?

## 3. Choose thresholds on validation data — 9 minutes

Predict default when the fitted probability is **at least the threshold**. For this lab, choose each threshold to maximize **validation balanced accuracy**:

$$
\text{Balanced accuracy} = \frac{\text{Sensitivity} + \text{Specificity}}{2}.
$$

This weights the two class-specific success rates equally, which is useful here because defaults are rare. It is our common teaching criterion; a real application may use different error costs. Keep this criterion fixed for all three models.

`pROC::coords(..., x = "best", best.method = "youden")` finds thresholds that maximize sensitivity plus specificity, so it also maximizes balanced accuracy. The probabilities and outcomes supplied to the ROC objects below are **validation** values. If several cutoffs tie, we use the smallest tied cutoff, favoring sensitivity.

```r
p.valid.logit <- predict(logit.fit, newdata = valid, type = "response")
p.valid.probit <- predict(probit.fit, newdata = valid, type = "response")
p.valid.cloglog <- predict(cloglog.fit, newdata = valid, type = "response")

# Higher predicted probabilities indicate default under all three links.
roc.valid.logit <- roc(valid$default, p.valid.logit,
                       levels = c("No", "Yes"), direction = "<", quiet = TRUE)
roc.valid.probit <- roc(valid$default, p.valid.probit,
                        levels = c("No", "Yes"), direction = "<", quiet = TRUE)
roc.valid.cloglog <- roc(valid$default, p.valid.cloglog,
                         levels = c("No", "Yes"), direction = "<", quiet = TRUE)

best.logit <- pROC::coords(roc.valid.logit, x = "best", best.method = "youden",
  ret = c("threshold", "sensitivity", "specificity"), transpose = FALSE)
best.probit <- pROC::coords(roc.valid.probit, x = "best", best.method = "youden",
  ret = c("threshold", "sensitivity", "specificity"), transpose = FALSE)
best.cloglog <- pROC::coords(roc.valid.cloglog, x = "best", best.method = "youden",
  ret = c("threshold", "sensitivity", "specificity"), transpose = FALSE)

# Resolve ties using the rule stated before looking at test results.
best.logit <- best.logit[which.min(best.logit$threshold), ]
best.probit <- best.probit[which.min(best.probit$threshold), ]
best.cloglog <- best.cloglog[which.min(best.cloglog$threshold), ]

t.logit <- best.logit$threshold
t.probit <- best.probit$threshold
t.cloglog <- best.cloglog$threshold

validation.results <- rbind(Logit = best.logit, Probit = best.probit,
                            Cloglog = best.cloglog)
validation.results$BalancedAccuracy <-
  (validation.results$sensitivity + validation.results$specificity) / 2
round(validation.results, 4)
```

**Before continuing:** Record all three thresholds and identify the link with the highest validation balanced accuracy, noting any tie. Why might the links choose different thresholds? Why is the largest validation score not an independent assessment of the chosen rule?

**Freeze the three fitted models and thresholds now.** Keep their full numerical precision; round only the displayed results. For this lab, do not refit the models on training plus validation data, because that would change the fitted rules whose thresholds you just selected.

## 4. Compare the fixed rules on test data — 6 minutes

Use the test set for final evaluation. The following code uses the saved thresholds; it does **not** search for new ones. Report all three models even if their ranking differs from validation.

```r
p.test.logit <- predict(logit.fit, newdata = test, type = "response")
p.test.probit <- predict(probit.fit, newdata = test, type = "response")
p.test.cloglog <- predict(cloglog.fit, newdata = test, type = "response")

# Confusion matrices: predictions are rows; actual outcomes are columns.
table(Predicted = factor(ifelse(p.test.logit >= t.logit, "Yes", "No"),
                         levels = c("No", "Yes")), Actual = test$default)
table(Predicted = factor(ifelse(p.test.probit >= t.probit, "Yes", "No"),
                         levels = c("No", "Yes")), Actual = test$default)
table(Predicted = factor(ifelse(p.test.cloglog >= t.cloglog, "Yes", "No"),
                         levels = c("No", "Yes")), Actual = test$default)

roc.test.logit <- roc(test$default, p.test.logit,
                      levels = c("No", "Yes"), direction = "<", quiet = TRUE)
roc.test.probit <- roc(test$default, p.test.probit,
                       levels = c("No", "Yes"), direction = "<", quiet = TRUE)
roc.test.cloglog <- roc(test$default, p.test.cloglog,
                        levels = c("No", "Yes"), direction = "<", quiet = TRUE)

# Numeric x means evaluate this FIXED threshold; x = "best" is not used here.
test.logit <- pROC::coords(roc.test.logit, x = t.logit, input = "threshold",
  ret = c("threshold", "sensitivity", "specificity", "accuracy"), transpose = FALSE)
test.probit <- pROC::coords(roc.test.probit, x = t.probit, input = "threshold",
  ret = c("threshold", "sensitivity", "specificity", "accuracy"), transpose = FALSE)
test.cloglog <- pROC::coords(roc.test.cloglog, x = t.cloglog, input = "threshold",
  ret = c("threshold", "sensitivity", "specificity", "accuracy"), transpose = FALSE)

test.results <- rbind(Logit = test.logit, Probit = test.probit,
                      Cloglog = test.cloglog)
test.results$BalancedAccuracy <-
  (test.results$sensitivity + test.results$specificity) / 2
test.results$AUC <- c(as.numeric(auc(roc.test.logit)),
                     as.numeric(auc(roc.test.probit)),
                     as.numeric(auc(roc.test.cloglog)))
round(test.results, 4)
```

**Check one row by hand:** Use its confusion matrix to calculate sensitivity and specificity. Confirm that their average equals the reported balanced accuracy. AUC summarizes ranking across thresholds; it is not the accuracy at your chosen cutoff.

## 5. Submit and compare across groups — 5 minutes

Submit **one Markdown response per group** where the instructor posts this lab. Include your group number, seed (`2400 + group`), validation-selected link or tie, and this table. Report four decimal places.

| Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC |
|---|---|---|---|---|---|---|---|
| Logit | | | | | | | |
| Probit | | | | | | | |
| Cloglog | | | | | | | |

Add brief answers:

1. Which link has the highest test balanced accuracy? Does the test AUC ranking agree? Describe the size of the differences.
2. Explain why using `x = "best"` on a **test** ROC object and then reporting the resulting test sensitivity and specificity as independent performance would double dip the data.
3. Compare your link rankings with one other group. What changed, and what does this suggest about declaring one link universally best? The groups' data subsets overlap, so these are examples of split variation, not six independent replications.

Do not change the seed, formula, threshold criterion, or thresholds after seeing test results. Our goal is to compare the three prespecified procedures honestly, not to improve their reported test scores.

Reference for the threshold calculation: [pROC `coords()` documentation](https://search.r-project.org/CRAN/refmans/pROC/html/coords.html).

**Comment by jmames17** [comment]
Group 2: @jmames17 @cjdow
| Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC |
|-- | -- | -- | -- | -- | -- | -- | -- |
| Logit | 0.0385  |  0.8764 |  0.9394 |  0.8718 |  0.874 | 0.9056  | 0.9610 |
| Probit | 0.0456  | 0.8754  | 0.9242  | 0.8692  | 0.871  | 0.8967  | 0.9609 | 
| Cloglog | 0.0408  | 0.8782  | 0.9394  | 0.8728  | 0.875  |  0.9061 | 0.9612 |  

1. The complementary log-log link has the highest test balanced accuracy at 0.9061. The test AUC ranking does agree, but it is higher at 0.9612. The test AUC is higher because it evaluates ranking  across thresholds rather than the single threshold that balanced accuracy is evaluating.
2. This is double-dipping because we are using data to evaluate which link is best and the reusing that data to evaluate its performance on an "independent" set of data.

**Comment by jcarp333** [comment]
Group3: @Peyton-Math, @jcarp333, @zhuang-jy
Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC
-- | -- | -- | -- | -- | -- | -- | --
Logit | 0.0349  |  0.8728 | 0.8788  | 0.8728  |  0.873 | 0.8758  | 0.9533  |
Probit | 0.044  | 0.8741  | 0.8636  |  0.8775 | 0.877  |  0.8705 | 0.9533  |
Cloglog |  0.0367 |  0.8728 |  0.8788 | 0.8607  |  0.871 | 0.8748 | 0.9533 |

1. Logistic has the highest test balanced accuracy, but all the links have the same AUC.
2. x = "best" maximizes the sensitivity and specificity in the test data so reporting them overstates how if will affect other data.

**Comment by blurissa** [comment]
|Link|Validation Threshold| Validation balanced accuracy| Test sensitivity| Tets specificity | Test Accuracy| Test balanced accuracy| Test AUC|
| -- | -- | -- | -- | -- | -- | -- | -- |
Logit |  0.0362 | 0.9032| 0.8333 |  0.8697 | 0.8685  | 0.8515   | 0.9317  |  
Probit | 0.0421  | 0.9029 | 0.8333  |  0.8666 | 0.8655  |  0.8500 |  0.9313 |  
Cloglog |  0.0387 |  0.9027  | 0.8182 | 0.8707  |  0.8690 |  0.8445 | 0.9318  |  

1. The link that has the highest test balanced accuracy is the Logit, by a very small margin. This does not agree with the the best AUC ranking, which is Cloglog, which is even smaller. 
2. This is double-dipping because we would be reusing the data to test link performance, and then using that data again to see its performance on non-independent data since we used the maximized sensitivity and specificity.
3.

**Comment by jess-l-b** [comment]
group 4. seed = 2404

Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC
-- | -- | -- | -- | -- | -- | -- | --
Logit | 0.0227  | 0.8658 | 0.9545 | 0.8340 | 0.8380 | 0.8943 | 0.9611  |  
Probit | 0.0244 | 0.8652 | 0.9545 | 0.8325 | 0.8365 | 0.8935 | 0.9610 |  
Cloglog | 0.0251 | 0.8658 | 0.9545 | 0.8340 | 0.8380 | 0.8943 | 0.9614 |

1. Logit and Cloglog both have the highest balanced accuracy, but Cloglog has the highest AUC.
2. Using the same data to set and evaluate the threshold may overestimate its sensitivity and specificity because the threshold is determined by maximizing both.
3. According to our test AUC, Cloglog is the best model while Probit is the worst. However, group

**Comment by Isabellhaines** [comment]
Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC
-- | -- | -- | -- | -- | -- | -- | --
Logit |  0.0361 |  0.8954 |  0.8788 |  0.8821 |  0.882 |  0.8804 |  0.9508
Probit |  0.0447 |  0.8954 | 0.8788  | 0.8821  |  0.882 | 0.8804  |  0.9506
Cloglog |  0.0359 | 0.8949  | 0.8788  |  0.8811 | 0.881  | 0.8799  |  0.9507

<br class="Apple-interchange-newline">

Add brief answers:

1. Logit and Probit have the highest test balance accuracy. The test AUC agrees with the test balance accuracy because both numbers show that the classifier is good at distinguishing the model. The differences between are very small which means any of the links could be good options.
2. Using x = "best" and reporting the resulting test sensitivity and specificity as independent performance would double dip the data because it maximizes sensitivity plus specificity, so it also maximizes balanced accuracy. This overstates how x effects the data.

**Comment by nmayers517** [comment]
Group 6:

<!DOCTYPE html>
Link | Validation threshold | Validation balanced accuracy | Test sensitivity | Test specificity | Test accuracy | Test balanced accuracy | Test AUC
-- | -- | -- | -- | -- | -- | -- | --
Logit | 0.0361  | 0.8954  |  0.8788 | 0.8821  | 0.882  | 0.8804  |  0.9508
Probit | 0.0447  |  0.8954 |  0.8788 | 0.8821  | 0.882  |  0.8804  |  0.9506
Cloglog | 0.0359  |  0.8949 |  0.8788 | 0.8811  |  0.881 |  0.8799 | 0.9507

<br class="Apple-interchange-newline">


1. The Logit and Probit tests are equal with 0.8804 Test Balanced Accuracy. The AUC ranking agrees with the Test Balanced Accuracy, with the links having Test AUCs of 0.9508 and 0.9506 respectively. The differences are minimal, only having differences to the fourth decimal point, indicating only small difference between links.
2. Using x = "best" double dips the data because it tries to maximize sensitivity and specificity at the same time, meaning that they were not independently taken. Claiming so overstates the data.
